[MM-9904] Add /invite slash command to invite users to a channel

cpanato opened the pull request:
#### Summary
 Add /invite slash command to invite users to a channel

#### Ticket Link
Jira: https://mattermost.atlassian.net/browse/MM-9904
GitHub: https://github.com/mattermost/mattermost-server/issues/8476

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json) updates

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-08a6a0934b7d8cf14.spinmint.com

Test Admin Account: Email: `user-0@sample.mattermost.com` | Password: `user-0`

Test User Account: Email: `user-1@sample.mattermost.com` | Password: `user-1`

Instance ID: i-08a6a0934b7d8cf14

amyblais commented:
Hi @cpanato,

Thank you for this contribution!

I tested this and I have a few small comments:
1. There is a small typo in the /invite help text - it should say "Invite a user to a channel" (“a user” not “an user”).
2. Maybe emphasize in the /invite help text that we should add the @-sign with the username (for clarity).
3. I'm able to add an existing user to a channel - is this expected or would an error message be better (something like "User is already in the channel"). 0/5

Possible bug:

Repro steps:
After I add a user in a channel they appear as “online” in the channel members drop-down list even though I’m not logged in as that user. After I add another user to the same channel, the previously added user becomes “offline”. I would expect that the added user always appears "offline" if I'm not logged in as that user. Doesn't repro on mobile, just on webapp and desktop app.

cpanato commented:
@amyblais thanks for reviewing it 😄 
1 -fixed
2 - I changed the text. added the `@` please let me know if that is enough.
3 - fixed as well, if the user is already in the channel it post a message saying the user is already in the channel

for the bug, also fixed.

do you mind to do another round of validation?

thanks

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-027fafb9953695e14.spinmint.com

Test Admin Account: Email: `user-0@sample.mattermost.com` | Password: `user-0`

Test User Account: Email: `user-1@sample.mattermost.com` | Password: `user-1`

Instance ID: i-027fafb9953695e14

amyblais commented:
Hi @cpanato,

Thank you for the changes, looks good! One other thing that I noticed is that our existing slash command for `/invite_people` is missing, wondering if you can help add it back?

cpanato commented:
@amyblais you need to enable the notification by email. if the email is disabled this slash command will not appear.

jasonblais commented:
Thanks @cpanato and @amyblais! Overall this looks great, just a few tweaks on messages.

1. When the user is added to the channel via `/invite`, two messages are posted. Propose we don't post the ephemeral message as the information is duplicated.

![image](https://user-images.githubusercontent.com/13119842/37789185-a4893740-2dd9-11e8-92e9-46b195a6f53d.png)

2. When a user isn't found, propose we use the same text as when the user isn't found for `/kick` or `/remove` commands, for consistency.

Observed:

![image](https://user-images.githubusercontent.com/13119842/37788165-1e88eb74-2dd7-11e8-8490-a80747c22164.png)

Expected (when using `/kick` or `/remove` commands):

![image](https://user-images.githubusercontent.com/13119842/37788171-23ea8e42-2dd7-11e8-8902-31f138c55ed9.png)

3. As @amyblais proposed, when a user already exists in the channel, propose we have the message `@[user] is already in the channel.`

Observed (not clear who `User` is)

![image](https://user-images.githubusercontent.com/13119842/37789328-fe5591a6-2dd9-11e8-8eb9-658a7bf63a0b.png)

4. Wrong message when `/invite` is used in direct message channels

Observed:
![image](https://user-images.githubusercontent.com/13119842/37789425-3a14e3ae-2dda-11e8-9c7e-1c0edf7ad011.png)

Expected: Text reads `You can't add someone to a direct message channel.` 

(Consistent with text when trying to use `remove` on a direct message channel)

5. When listing users in the `/invite @` user autocomplete, it lists channel members first. This can be a separate improvement, but it would be nicer to only list users not in the channel given those are the users you're trying to add. It could be same results returned in the "Add Members" dialog.

Observed:
![image](https://user-images.githubusercontent.com/13119842/37789547-939bd31a-2dda-11e8-8abc-1907d52f0980.png)

Expected:
![image](https://user-images.githubusercontent.com/13119842/37789562-9f47e79e-2dda-11e8-8b16-2927181a7c66.png)

esethna commented:
@cpanato @jasonblais Thanks for your work on this PR! 

One piece of additional feedback: wondering if it would be possible to accept channel names with a tild (~channelname)? So people can use the channel autocomplete if they are not inviting to the current channel (in which case you don't need the channel name). 

So for example, the command would accept an input such as `/invite @userA ~channelX`

cpanato commented:
@jasonblais @esethna 
I did the changes requested. 
for the item 5, I think will be better another ticket.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-003a926b47f0e2d8c.spinmint.com

Test Admin Account: Email: `user-0@sample.mattermost.com` | Password: `user-0`

Test User Account: Email: `user-1@sample.mattermost.com` | Password: `user-1`

Instance ID: i-003a926b47f0e2d8c

jasonblais commented:
Thanks! Re-tested the following notes + Eric's. See notes on 1, 3 and 4.

1 - This is now working - however I get an error each time I try to add a user to a channel I'm not currently viewing

![image](https://user-images.githubusercontent.com/13119842/37904084-8878c7ea-30c8-11e8-8dcb-b4a7823f1488.png)

2 - Fixed

3 - Looks like you added the channel name to the system message. That's okay, except it's using the channel handle which could be confusing (in the attached, I typed `/invite @[user] autem` and get a system message `brandon.ford is already in the ratione-1.` since the channel handle and display name do not match).

![image](https://user-images.githubusercontent.com/13119842/37904148-cd5bdf5a-30c8-11e8-8bac-0d77b5c23e5c.png)
`/

Propose we keep `@[user] is already in the channel.` for simplicity.

4 - Still an issue for me

![image](https://user-images.githubusercontent.com/13119842/37904210-07d9d2e0-30c9-11e8-8674-0aa73861b7b4.png)

5) Given this command can be used to add people to a channel you're not currently viewing, this proposed change won't actually work. So I won't create a separate ticket for it now.

6) (Eric's note) tested and works.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-00c0483f6915768fe.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-00c0483f6915768fe

cpanato commented:
Lets sync in the pre-release @jasonblais

cpanato commented:
@jasonblais did some changes. If you still have issues with item 1, lets sync in the pre-release.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-028f6e12ef28fa2a1.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-028f6e12ef28fa2a1

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-059b400b650a14563.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-059b400b650a14563

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-06b20f2c0a796ec3e.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-06b20f2c0a796ec3e

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-00391c59084ef9ee0.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-00391c59084ef9ee0

jasonblais commented:
@cpanato Minor updates made to UI strings. Awaiting for the permissions fix, after which we can rebase and test this PR.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0037d65547eba5b0d.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0037d65547eba5b0d

jasonblais commented:
@cpanato Sorry, still having trouble where it cannot find the channel by its channel handle.

![image](https://user-images.githubusercontent.com/13119842/38496577-1cf22480-3bcc-11e8-9573-8593335fe1a3.png)

Note that it works for `/mute` okay:

![image](https://user-images.githubusercontent.com/13119842/38496609-3686999e-3bcc-11e8-93ee-dab94af6565f.png)

PS: There's also an additional space between the channel name and the period. Can you help take a look?

![image](https://user-images.githubusercontent.com/13119842/38496653-50beb15c-3bcc-11e8-8f0c-0dc60d1103c6.png)

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-026a48d2f1691c105.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-026a48d2f1691c105

jespino commented on app/command_invite.go:
```diff
@@ -0,0 +1,113 @@
+			targetChannelName = splitMessage[1][1:]
+		} else {
+			targetChannelName = splitMessage[1]
+		}
```
Can we use TrimPrefix here to (like in targetUsername)?

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0ba06ca0c339e5705.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0ba06ca0c339e5705

jasonblais commented:
@cpanato If you can remove any code added for testing/debugging, I'll queue this for dev review.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0c45833079a160ea3.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0c45833079a160ea3

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-084e1912f9671c9a7.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-084e1912f9671c9a7

mattermod commented:
Spinmint test server destroyed

jespino commented on app/command_invite.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	// Check if user is already in the channel
+	userChannels, err := a.GetChannelsForUser(args.TeamId, userProfile.Id)
```
Why not check directly for membership existence? instead of loop over all the user channels something like `a.GetChannelMember(channelToJoin.Id, userProfile.Id); if err == nil { return error}`.

cpanato replied:
good tip. I was looking for something, but did not think about this one 😄  thanks so much! will update that

cpanato replied:
Done. PTAL

jespino commented on app/command_invite.go:
```diff
@@ -0,0 +1,104 @@
+		}
+	}
+
+	if channelToJoin.Type == model.CHANNEL_PRIVATE && !a.SessionHasPermissionToChannel(args.Session, channelToJoin.Id, model.PERMISSION_READ_CHANNEL) {
```
The permissions checking would be the same that used in `api4/channel.go` `addChannelMember` function. It is a bit more complex (depends on if you are adding yourself, or other, and if you have the permissions to manage the team members).

the code in the other file is:
```go
    if channel.Type == model.CHANNEL_OPEN {
        if member.UserId == c.Session.UserId {
            if !c.App.SessionHasPermissionToChannel(c.Session, channel.Id, model.PERMISSION_JOIN_PUBLIC_CHANNELS) {
                c.SetPermissionError(model.PERMISSION_JOIN_PUBLIC_CHANNELS)
                return
            }
        } else {
            if !c.App.SessionHasPermissionToChannel(c.Session, channel.Id, model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS) {
                c.SetPermissionError(model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS)
                return
            }
        }
    }

    if channel.Type == model.CHANNEL_PRIVATE && !c.App.SessionHasPermissionToChannel(c.Session, channel.Id, model.PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS) {
        c.SetPermissionError(model.PERMISSION_MANAGE_PRIVATE_CHANNEL_MEMBERS)
        return
    }
```

Maybe in this case we can assume that you are trying to invite another person, so you will never invite yourself to channels.

cpanato replied:
done! thanks so much for your review. PTAL

saturninoabril commented on app/command_invite.go:
```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2018-present Mattermost, Inc. All Rights Reserved.
```
I think the latest agreement is just to set this to "2015-present"

cpanato replied:
ahh. I did not know that. will update. thanks for the review!

cpanato replied:
done
